### PR TITLE
Document error message and documentation conventions in markdown file

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
## Description of changes

Aggregate content from an internal wiki and discussion on #745 to document error and docs conventions. Some minor tweaks based on my understanding of our current practice.

Please comment with any other code style conventions we follow and should document. I probably won't incorporate them into this PR unless they directly relate to errors/docs, but it'd be good to get anything else written down.  

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [x] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar language specification.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in [`cedar-docs`](https://github.com/cedar-policy/cedar-docs). PRs should be targeted at a `staging-X.Y` branch, not `main`.)
- [ ] I'm not sure how my change impacts the documentation. (Post your PR anyways, and we'll discuss in the comments.)

### CODE REVIEW

File: style_guide.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+## Conventions for Cedar Documentation
+
+Everything publicly exported from `cedar-policy` (functions, types, modules, etc.) must be documented with a doc comment (a comment delimited by `///`).
+There is no minimum standard for length of documentation. A comment be as short as a single sentence fragment.
+
+**Capitalization:** Doc comments (sentence fragments or otherwise) should always begin with a capital letter.
+
+**Punctuation:** Sentence fragments (e.g., ``Errors that can occur when adding to a `PolicySet` ``) do not need a period, but 
+proper sentences (e.g. ``There was a duplicate `PolicyId` encountered in either the set of templates or the set of policies.``) do. 
+Use periods in the case of a mixture of sentence fragments and full sentences, use periods.
```
Comment (khieta) [inline style_guide.md:18]:
```suggestion
In the case of a mixture of sentence fragments and full sentences, use periods.
```

File: style_guide.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+These invariants should not generally be included in doc comments, i.e., they should start with `//` instead of `///`.
+
+**Other:** When describing an identifier, use “id”, not “Id” or “ID” (applies to both doc strings and error messages).
+
+## Conventions for Cedar Errors
+
+### Error Types
+
+Methods that can return errors return a `Result` type where the error type is a public enum. 
+Each variant of that enum is either another error enum or a single struct with private field(s).
```
Comment (khieta) [inline style_guide.md:30]:
```suggestion
Methods that can return errors should return a `Result` type where the error type is a public enum. 
Each variant of that enum should be either another error enum or a single struct with private field(s). 
```

File: style_guide.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+Methods that can return errors return a `Result` type where the error type is a public enum. 
+Each variant of that enum is either another error enum or a single struct with private field(s). 
+The structs for a particular error enum are typically gathered into a submodule.
+The names of all error enums and structs should end in `Error`, but each variant of an error enum should not end in `Error`.
+
+Error enums are often annotated with `#[non_exhaustive]`, though we omit this annotation when we feel callers _should_ have their build break when we introduce new error variants.
+For example, `#[non_exhaustive]` should never be used for `EvaluationError` or other errors that are returned from the `is_authorized` API, but parsing and validation errors generally should be `#[non_exhaustive]`.
+
+The enum itself, and also each of the member structs, all implement `miette::Diagnostic` and `thiserror::Error`.
```
Comment (khieta) [inline style_guide.md:37]:
```suggestion
The enum itself, and also each of the member structs, should implement `miette::Diagnostic` and `thiserror::Error`.
```

File: style_guide.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+Both of these should be automatically derived when possible, and the `#[error(...)]` macro should be used to generate `Display` implementations.
+For more complex error it is occasionally easier to manually implement `Display`.
+Manual implementation of `miette::Diagnostic` is less common, but is occasionally necessary to work around limitations in derive macro.
+We use a [module of internal macros](https://github.com/cedar-policy/cedar/blob/main/cedar-policy-core/src/error_macros.rs) to help in these situations.
+
+By ensuring the internal structs have private fields we are free to track more details in the error without making a breaking change to our API.
```
Comment (khieta) [inline style_guide.md:43]:
```suggestion
By ensuring the internal structs have private fields we are free to change details in the error without making a breaking change to our API.
```

File: style_guide.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+### Error Documentation
+
+Error type documentation follows the same conventions as documentation on other types.
+Each error needs to be documented twice (once on the enum variant and again on the struct).
+To avoid duplicating large comments we prefer to place detailed documentation on the enum variant with a shorter comment on the struct.
+
+### Error Messages
+
+* Single-sentence error messages should not use capitalization and periods.
```
Comment (khieta) [inline style_guide.md:93]:
```suggestion
* Single-sentence error messages should not use capitalization or periods.
```

File: style_guide.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  These should independently follow the recommendations for error message punctuation and capitalization.
+* User-provided input (e.g. attributes, entity uids, policy ids) should be surrounded with backticks or be placed after a colon at the end of the message.
+  Use of backticks is preferred.
+  Examples: ``undeclared action `foo` ``, `undeclared entity type(s): {"Foo"}`
+* Whenever possible, include a [source span](https://docs.rs/miette/latest/miette/#-snippets) indicating where in the input the error occurred. Consider if this span adequately replaces any verbatim copies of user-provided input included in the error message.
+* Related error messages should be concatenated with colons. Example: ``error occurred while evaluating policy `policy0`: `User::"alive"` does not have the attribute `foo` ``"
+* Error messages should not refer to specific Rust type names (like PolicyId), preferring to refer to generic Cedar concepts (like policy id). Remember that these errors will be viewed by Cedar users who may never interact directly with the Rust library.
```
Comment (khieta) [inline style_guide.md:103]:
```suggestion
* Error messages should not refer to specific Rust type names (like `PolicyId`), preferring to refer to generic Cedar concepts (like policy id). Remember that these errors will be viewed by Cedar users who may never interact directly with the Rust library.
```

**Review by cdisselkoen — APPROVED** [review]
Looks great

**Review by aaronjeline — APPROVED** [review]

**Pull request merged**